### PR TITLE
HDDS-7087. Manage RocksObjects to detect leaks

### PR DESCRIPTION
**Pull request by duongnguyen0**
## What changes were proposed in this pull request?
We need a mechanism to detect if any RocksObject is not closed before being GCed. 
This PR replaces the usage of RocksObject classes in `org.rocksdb.*` package by managed RocksObject classes whose finalizers are intercepted to assert of their instances are closed before GCed.

https://issues.apache.org/jira/browse/HDDS-7087

### Metric showing memory leaks.
Screen Shot 2022-08-04 at 11 11 08 AM

### Sample error message when RocksObject is used in Ozone.
```
Banned imports detected:

Reason: Use managed RocksObjects under org.apache.hadoop.hdds.utils.db.managed instead.
	in file: org/apache/hadoop/hdds/utils/db/DBStoreBuilder.java
		org.rocksdb.RocksDB                   	(Line: 52, Matched by: org.rocksdb.**)
	in file: org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java
		org.rocksdb.WriteBatch                	(Line: 22, Matched by: org.rocksdb.**)
		org.rocksdb.WriteOptions              	(Line: 23, Matched by: org.rocksdb.**)
	in file: org/apache/hadoop/hdds/utils/db/TableConfig.java
		org.rocksdb.ColumnFamilyOptions       	(Line: 28, Matched by: org.rocksdb.**)
	in file: org/apache/hadoop/hdds/utils/db/RDBSstFileWriter.java
		org.rocksdb.EnvOptions                	(Line: 23, Matched by: org.rocksdb.**)
		org.rocksdb.Options                   	(Line: 24, Matched by: org.rocksdb.**)
		org.rocksdb.SstFileWriter             	(Line: 26, Matched by: org.rocksdb.**)
	in file: org/apache/hadoop/hdds/utils/db/RDBSstFileLoader.java
		org.rocksdb.IngestExternalFileOptions 	(Line: 20, Matched by: org.rocksdb.**)
	in file: org/apache/hadoop/hdds/utils/db/RDBStore.java
		org.rocksdb.DBOptions                 	(Line: 41, Matched by: org.rocksdb.**)
		org.rocksdb.TransactionLogIterator    	(Line: 43, Matched by: org.rocksdb.**)
```

## How was this patch tested?

Standard CI: https://github.com/duongnguyen0/ozone/actions/runs/2805369172

### CODE REVIEW

**Comment by kerneltime** [comment]
I like this approach; @jojochuang mentioned that there might be ways in the pom for all packages other than managed packaged to block rocks DB import.

**Comment by jojochuang** [comment]
There's a Maven plugin to ban certain classpath: https://maven.apache.org/enforcer/enforcer-rules/bannedDependencies.html
example: https://github.com/apache/hadoop/pull/2969/files

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksIterator.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Managed RocksIterator.
+ */
+public class ManagedRocksIterator extends RocksIterator {
+  private final RocksIterator original;
+
+  public ManagedRocksIterator(RocksDB parent, RocksIterator original) {
+    // This object just act as a proxy to the real iterator.
+    super(parent, original.getNativeHandle());
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksIterator.java:34]: There are two RocksIterator objects which may cause problems in the future.  E.g. RocksDB may add a new method to RocksIterator but it is not overridden in ManagedRocksIterator.  I suggest to have a get() method as below.
```
public class ManagedRocksIterator {
  private final RocksIterator original;

  public ManagedRocksIterator(RocksIterator original) {
    this.original = original;
  }

  public RocksIterator get() {
    return original;
  }

  @Override
  protected void finalize() throws Throwable {
    ManagedRocksObjectUtils.assertClosed(original);
    super.finalize();
  }
}
```

**Review by szetszwo — COMMENTED** [review]
@duongnguyen0 , thanks a lot for working on this!  The idea of having managed objects is great.  We should remove the import of org.rockdb.* from the other classes in order to avoid missing the changes.  See also the comments inlined.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/AbstractDatanodeStore.java [hunk 2]
```diff
@@ -116,7 +117,7 @@ public void start(ConfigurationSource config)
               OZONE_METADATA_STORE_ROCKSDB_STATISTICS_DEFAULT);
 
       if (!rocksDbStat.equals(OZONE_METADATA_STORE_ROCKSDB_STATISTICS_OFF)) {
-        Statistics statistics = new Statistics();
+        Statistics statistics = new ManagedStatistics();
```
[the remark this answers is not in the document]
Comment (duongnguyen0) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/AbstractDatanodeStore.java]: Updated.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksIterator.java (hunk 1 above)
Comment (duongnguyen0) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedRocksIterator.java]: Good one, this make the code a lot simpler actually.

**Comment by duongnguyen0** [comment]
Thanks @kerneltime, @jojochuang and @szetszwo for the early review and suggestions. I've updated the code to (mostly) ban the usage of `org.rocksdb`.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

Just a minor comment inlined.

**Comment by szetszwo** [comment]
> 2 successful and 8 skipped checks

Why the checks were skipped?

**Comment by duongnguyen0** [comment]
> > 2 successful and 8 skipped checks
> 
> Why the checks were skipped?

Because I was a draft PR. Now all the checks are running.

**Pull request merged**